### PR TITLE
[MOV] web: Moved CodeEditor from studio to web

### PR DESCRIPTION
**Pull request by BastienFafchamps**
This component was created as an easy way to add a code editing field
without needing a record (unlike the AceField component). It also
completely abstracts the use of the Ace Editor library.

The name "CodeEditor" was purposely chosen to clearly state what the
component is used for and abstract the use of Ace.

Enterprise PR: https://github.com/odoo/enterprise/pull/39365

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web/static/src/core/code_editor/code_editor.js [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        },
+    };
+    static defaultProps = {
+        value: "",
+        onChange: () => {},
+        class: "",
+        theme: "",
+    };
+
+    static ACE_LIBS = ["/web/static/lib/ace/ace.js"];
```
Comment (ged-odoo) [inline addons/web/static/src/core/code_editor/code_editor.js:49]: i think we should create a bundle with all ace files mentioned here, that would be simpler

File: addons/web/static/src/core/code_editor/code_editor.js [hunk 2]
```diff
@@ -0,0 +1,137 @@
+/** @odoo-module */
+import { Component, onWillDestroy, onWillStart, useEffect, useRef } from "@odoo/owl";
+import { loadBundle, loadJS } from "@web/core/assets";
+import { useDebounced } from "@web/core/utils/timing";
+
+function onResized(ref, callback) {
+    const _ref = typeof ref === "string" ? useRef(ref) : ref;
+    const resizeObserver = new ResizeObserver(() => callback());
```
Comment (ged-odoo) [inline addons/web/static/src/core/code_editor/code_editor.js:8]:
```suggestion
    const resizeObserver = new ResizeObserver(callback);
```

File: addons/web/static/src/core/code_editor/code_editor.xml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<templates>
+
+    <t t-name="web.CodeEditor" owl="1">
+        <div t-ref="editorRef" id="ace-editor" class="o_web_code_editor w-100 h-100" t-att-class="props.class">
```
Comment (ged-odoo) [inline addons/web/static/src/core/code_editor/code_editor.xml:5]: can you remove the `id and `o_web_code_editor` class?

File: addons/web/static/src/core/code_editor/code_editor.js (hunk 1 above)
Comment (BastienFafchamps) [reply addons/web/static/src/core/code_editor/code_editor.js:49]: Something like that or a python bundle?
```js
onWillStart(() => {
    const aceBundle = {
        jsLibs: [
            "/web/static/lib/ace/ace.js",
            [
                "/web/static/lib/ace/mode-js.js",
                "/web/static/lib/ace/javascript_highlight_rules.js",
                "/web/static/lib/ace/mode-xml.js",
                "/web/static/lib/ace/mode-qweb.js",
                "/web/static/lib/ace/mode-scss.js",
                "/web/static/lib/ace/theme-monokai.js",
            ]
        ]
    }
    return loadBundle(aceBundle);
});
```

File: addons/web/static/src/core/code_editor/code_editor.js [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    );
+
+    onWillDestroy(() => {
+        resizeObserver.disconnect();
+    });
+}
+
+export class CodeEditor extends Component {
+    static template = "web.CodeEditor";
+    static components = {};
```
Comment (ged-odoo) [inline addons/web/static/src/core/code_editor/code_editor.js:27]: not really necessary

**Comment by ged-odoo** [comment]
robodoo rebase-ff

**Comment by ged-odoo** [comment]
robodoo r+

**Pull request closed without merging**